x-pack/plugins/cases/public/components/all_cases/table_filter_config/use_filter_config.tsx

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useEffect, useState } from 'react';
+import useLocalStorage from 'react-use/lib/useLocalStorage';
+import { LOCAL_STORAGE_KEYS } from '../../../../common/constants';
+import type { FilterChangeHandler, FilterConfig, FilterConfigState } from './types';
+import { useCustomFieldsFilterConfig } from './use_custom_fields_filter_config';
+import { MoreFiltersSelectable } from './more_filters_selectable';
+import { useCasesContext } from '../../cases_context/use_cases_context';
+
+const mergeSystemAndCustomFieldConfigs = ({
+  systemFilterConfig,
+  customFieldsFilterConfig,
+}: {
+  systemFilterConfig: FilterConfig[];
+  customFieldsFilterConfig: FilterConfig[];
+}) => {
+  const newFilterConfig = new Map(
+    [...systemFilterConfig, ...customFieldsFilterConfig]
+      .filter((filter) => filter.isAvailable)
+      .map((filter) => [filter.key, filter])
+  );
+
+  return newFilterConfig;
+};
+
+export const useFilterConfig = ({
+  systemFilterConfig,
+  onFilterOptionChange,
+}: {
+  systemFilterConfig: FilterConfig[];
+  onFilterOptionChange: FilterChangeHandler;
+}) => {
+  const { appId } = useCasesContext();
+  const { customFieldsFilterConfig } = useCustomFieldsFilterConfig({ onFilterOptionChange });
+  const [filterConfigs, setFilterConfigs] = useState<Map<string, FilterConfig>>(
+    () => new Map([...systemFilterConfig].map((filter) => [filter.key, filter]))
+  );
+  const [activeByFilterKey, setActiveByFilterKey] = useLocalStorage<FilterConfigState[]>(
+    `${appId}.${LOCAL_STORAGE_KEYS.casesTableFiltersConfig}`,
+    []
+  );
+
+  useEffect(() => {
+    const newFilterConfig = mergeSystemAndCustomFieldConfigs({
+      systemFilterConfig,
+      customFieldsFilterConfig,
+    });
+
+    filterConfigs.forEach((filter) => {
+      if (!newFilterConfig.has(filter.key)) {
+        filter.deactivate();
+      }
+    });
+  }, [filterConfigs, systemFilterConfig, customFieldsFilterConfig]);
+
+  useEffect(() => {
+    setFilterConfigs(
+      mergeSystemAndCustomFieldConfigs({
+        systemFilterConfig,
+        customFieldsFilterConfig,
+      })
+    );
+  }, [systemFilterConfig, customFieldsFilterConfig]);
+
+  const onChange = ({ selectedOptionKeys }: { filterId: string; selectedOptionKeys: string[] }) => {
```

Comment:
Could you add some tests for all this logic in `use_filter_config.test.tsx`?

We can make sure different filters are active as a precondition, and then guarantee that calling `onChange`  with certain `selectedOptionKeys` calls `setActiveByFilterKey` with the right params.

Stuff like:
- testing the order is preserved
- `isActive=false` filters are ignored
- testing no longer available filters are cleaned up correctly
- new filters are appended

This method is so overwhelming 😄

Reply:
I'm testing it in the table so that the tests are less linked to implementation details like this hook

- testing the order is preserved [test](https://github.com/elastic/kibana/pull/171176/files#diff-827ef95cedb18b40d3dd8aaf5a64fc2cfb95b0adfbca807c0df6b64c543133efR675) [test](https://github.com/elastic/kibana/pull/171176/files#diff-827ef95cedb18b40d3dd8aaf5a64fc2cfb95b0adfbca807c0df6b64c543133efR712)
- isActive=false filters are ignored [test](https://github.com/elastic/kibana/pull/171176/files#diff-827ef95cedb18b40d3dd8aaf5a64fc2cfb95b0adfbca807c0df6b64c543133efR542)
- testing no longer available filters are cleaned up correctly [test](https://github.com/elastic/kibana/pull/171176/files#diff-827ef95cedb18b40d3dd8aaf5a64fc2cfb95b0adfbca807c0df6b64c543133efR416) [test](https://github.com/elastic/kibana/pull/171176/files#diff-827ef95cedb18b40d3dd8aaf5a64fc2cfb95b0adfbca807c0df6b64c543133efR560) [test](https://github.com/elastic/kibana/pull/171176/files#diff-827ef95cedb18b40d3dd8aaf5a64fc2cfb95b0adfbca807c0df6b64c543133efR767)
- new filters are appended [test](https://github.com/elastic/kibana/pull/171176/files#diff-827ef95cedb18b40d3dd8aaf5a64fc2cfb95b0adfbca807c0df6b64c543133efR489)